
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
        <link rel="next" href="assays/">
      
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.29">
    
    
      
        <title>Snailz Docs</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.76a95c52.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#snailz" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="Snailz Docs" class="md-header__button md-logo" aria-label="Snailz Docs" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Snailz Docs
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Snailz
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="Snailz Docs" class="md-nav__button md-logo" aria-label="Snailz Docs" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Snailz Docs
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Snailz
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="." class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Snailz
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#usage" class="md-nav__link">
    <span class="md-ellipsis">
      Usage
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#database" class="md-nav__link">
    <span class="md-ellipsis">
      Database
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-files" class="md-nav__link">
    <span class="md-ellipsis">
      Data Files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="assays/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assays
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="clui/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clui
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="db/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Db
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="genomes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Genomes
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="grid/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Grid
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="mangle/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Mangle
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="params/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Params
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="plates/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Plates
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="samples/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Samples
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="survey/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Survey
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#usage" class="md-nav__link">
    <span class="md-ellipsis">
      Usage
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#database" class="md-nav__link">
    <span class="md-ellipsis">
      Database
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-files" class="md-nav__link">
    <span class="md-ellipsis">
      Data Files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="snailz">Snailz</h1>
<p><img src="https://raw.githubusercontent.com/gvwilson/snailz/main/img/snail-logo.svg" alt="snail logo" width="200px"></p>
<p>These data generators model genomic analysis of snails in the Pacific Northwest
that are growing to unusual size as a result of exposure to pollution.</p>
<ul>
<li>One or more <em>surveys</em> are conducted at one or more <em>sites</em>.</li>
<li>Each survey collects <em>genomes</em> and <em>sizes</em> of snails.</li>
<li>A <em>grid</em> at each site is marked out to show the presence or absence of pollution.</li>
<li><em>Laboratory staff</em> perform <em>assays</em> of the snails' genetic material.</li>
<li>Each assay plate has a <em>design</em> showing the material applied and <em>readings</em> showing the measured response.</li>
<li>Plates may be <em>invalidated</em> after the fact if a staff member believes it is contaminated.</li>
</ul>
<p><img src="https://raw.githubusercontent.com/gvwilson/snailz/main/img/survey.png" alt="survey sites"></p>
<h2 id="usage">Usage</h2>
<ol>
<li>Create a fresh Python environment: <code>mamba create -y -n snailz python=3.12</code></li>
<li>Activate that environment: <code>mamba activate snailz</code></li>
<li>Build development version of package: <code>pip install -e .</code></li>
<li>View available commands: <code>snailz --help</code></li>
<li>Copy default parameter files: <code>snailz params --outdir .</code></li>
<li>See how to regenerate datasets: <code>python -c 'import snailz; help(snailz)'</code></li>
</ol>
<p>This project also includes a Makefile that will re-execute commands as needed.
To see available commands, run:</p>
<pre><code>make DATA=data PARAMS=snailz/params commands
</code></pre>
<p>(To keep the Makefile simple,
<code>DATA</code> and <code>PARAMS</code> must be defined even for commands that don't need them.)
To regenerate all of the datasets, run:</p>
<pre><code>make DATA=data PARAMS=snailz/params datasets
</code></pre>
<table>
<thead>
<tr>
<th>command</th>
<th>action</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>dev</code></td>
<td>rebuild development version of package</td>
</tr>
<tr>
<td><code>lint</code></td>
<td>check code using ruff</td>
</tr>
<tr>
<td><code>clean</code></td>
<td>remove datafiles</td>
</tr>
<tr>
<td><code>commands</code></td>
<td>show available commands</td>
</tr>
<tr>
<td><code>datasets</code></td>
<td>make all datasets</td>
</tr>
<tr>
<td><code>survey</code></td>
<td>generate survey map</td>
</tr>
<tr>
<td><code>mangled</code></td>
<td>create inconsistent plate files</td>
</tr>
<tr>
<td><code>db</code></td>
<td>generate database</td>
</tr>
<tr>
<td><code>plates</code></td>
<td>generate plate files</td>
</tr>
<tr>
<td><code>assays</code></td>
<td>generate assay files</td>
</tr>
<tr>
<td><code>samples</code></td>
<td>sample snails from survey sites</td>
</tr>
<tr>
<td><code>genomes</code></td>
<td>synthesize genomes</td>
</tr>
<tr>
<td><code>grids</code></td>
<td>synthesize pollution grids</td>
</tr>
</tbody>
</table>
<h2 id="database">Database</h2>
<p>The final database <code>data/lab.db</code> is structured as shown below.
Note that the data from the file <code>assays.json</code> is split between several tables.
Note also that the SQLite database file is <em>not</em> included in this repository
because its binary representation changes each time it is regenerated
(even though the values it contains stay the same).
The map of survey locations in <code>data/survey.png</code> is not included in the repository for the same reason,
but a duplicate is manually saved in <code>img/survey.png</code>.</p>
<p><img src="https://raw.githubusercontent.com/gvwilson/snailz/main/img/db-schema.svg" alt="database schema"></p>
<ul>
<li><code>site</code>: survey site<ul>
<li><code>site_id</code>: primary key (text)</li>
<li><code>lon</code>: longitude of site reference marker (float deg)</li>
<li><code>lat</code>: latitude of site reference marker (float deg)</li>
</ul>
</li>
<li><code>survey</code><ul>
<li><code>survey_id</code>: primary key (text)</li>
<li><code>site_id</code>: foreign key of site where survey was conducted (text)</li>
<li><code>date</code>: date that survey was conducted (date, YYYY-MM-DD)</li>
</ul>
</li>
<li><code>sample</code>: sample taken from survey<ul>
<li><code>sample_id</code>: primary key (int, 1-1 with <code>experiment.sample_id</code>)</li>
<li><code>survey_id</code>: foreign key of survey (int)</li>
<li><code>lon</code>: longitude of sample site (float deg)</li>
<li><code>lat</code>: latitude of sample site (float deg)</li>
<li><code>sequence</code>: genome sequence of sample (text)</li>
<li><code>size</code>: snail size (float)</li>
</ul>
</li>
<li><code>experiment</code>: experiment done on sample<ul>
<li><code>sample_id</code>: primary key (int, 1-1 with <code>sample.sample_id</code>)</li>
<li><code>kind</code>: kind of experiment (text, either 'ELISA' or 'JESS')</li>
<li><code>start</code>: start date (date, YYYY-MM-DD)</li>
<li><code>end</code>: end date (date, YYYY-MM-DD, null if experiment is ongoing)</li>
</ul>
</li>
<li><code>staff</code><ul>
<li><code>staff_id</code>: primary key (int)</li>
<li><code>personal</code>: personal name (text)</li>
<li><code>family</code>: family name (text)</li>
</ul>
</li>
<li><code>performed</code>: join table showing which staff members performed which experiments<ul>
<li><code>staff_id</code>: foreign key of staff member</li>
<li><code>sample_id</code>: foreign key of sample/experiment</li>
</ul>
</li>
<li><code>plate</code>: information about single assay plate<ul>
<li><code>plate_id</code>: primary key (int)</li>
<li><code>sample_id</code>: foreign key of sample/experiment (int)</li>
<li><code>date</code>: date that plate was run (date, YYYY-MM-DD)</li>
<li><code>filename</code>: filename of design/results file (text)</li>
</ul>
</li>
<li><code>invalidated</code>: invalidated plates<ul>
<li><code>plate_id</code>: foreign key of plate (int)</li>
<li><code>staff_id</code>: foreign key of staff member who did invalidation (int)</li>
<li><code>date</code>: when plate was invalidated</li>
</ul>
</li>
</ul>
<h2 id="data-files">Data Files</h2>
<p><code>./data</code> contains a generated dataset for reference.
As noted above,
it does <em>not</em> contain the SQLite database file <code>lab.db</code>;
run <code>snailz db</code> to regenerate it.
(See <code>help(snailz)</code> for an example invocation.)</p>
<ul>
<li>Genomes: <code>genomes.json</code><ul>
<li><code>length</code>: number of base pairs (int &gt; 0)</li>
<li><code>reference</code>: the unmutated reference genome (text)</li>
<li><code>individuals</code>: sequences for individuals (list of text)</li>
<li><code>locations</code>: locations of mutations (list of int)</li>
<li><code>susceptible_loc</code>: location of mutation of interest (int &gt;= 0)</li>
<li><code>susceptible_base</code>: mutated base responsible for size change (char)</li>
</ul>
</li>
<li>Grids: <code>grids/*.csv</code> (one file per site)<ul>
<li>1/0: presence/absence of contamination at sample location</li>
</ul>
</li>
<li>Samples: <code>grids/samples.csv</code><ul>
<li><code>sample_id</code>: unique ID for genetic sample (text)</li>
<li><code>survey_id</code>: which survey it was taken in (text)</li>
<li><code>lon</code>: longitude of sample site (float)</li>
<li><code>lat</code>: latitude of sample site (float)</li>
<li><code>sequence</code>: sampled gene sequence (text)</li>
<li><code>size</code>: snail weight (float, grams)</li>
</ul>
</li>
<li>Assays: <code>assays.json</code><ul>
<li><code>staff</code>:<ul>
<li><code>staff_id</code>: unique staff member identifier (int &gt; 0)</li>
<li><code>personal</code>: personal name (text)</li>
<li><code>family</code>: family name (text)</li>
</ul>
</li>
<li><code>experiment</code>: experiment details<ul>
<li><code>sample_id</code>: sample that experiment used (int &gt; 0)</li>
<li><code>kind</code>: "ELISA" or "JESS" (text)</li>
<li><code>start</code>: start date (date, YYYY-MM-DD)</li>
<li><code>end</code>: end date (date, YYYY-MM-DD or None if experiment incomplete)</li>
</ul>
</li>
<li><code>performed</code>: join table showing who performed which experiments<ul>
<li><code>staff_id</code>: foreign key to <code>staff</code></li>
<li><code>sample_id</code>: foreign key to <code>experiment</code></li>
</ul>
</li>
<li><code>plate</code>: details of assay plates used in experiments<ul>
<li><code>plate_id</code>: unique plate identifier (int &gt; 0)</li>
<li><code>sample_id</code>: foreign key to <code>sample</code> (text)</li>
<li><code>date</code>: date plate was run (date, YYYY-MM-DD)</li>
<li><code>filename</code>: name of design and results files (text)</li>
</ul>
</li>
<li><code>invalidated</code>: which plates have been invalidated<ul>
<li><code>plate_id</code>: foreign key to plate (text)</li>
<li><code>staff_id</code>: foreign key to staff member responsible (text)</li>
<li><code>date</code>: invalidation date (date, YYYY-MM-DD)</li>
</ul>
</li>
</ul>
</li>
<li>Plates are represented by matching files in the <code>designs</code> and <code>readings</code> directories<ul>
<li><code>designs/*.csv</code>: assay plate designs<ul>
<li>header: machine type, file type ("design" or "readings"), staff ID</li>
<li>blank line</li>
<li>table with column and row titles showing material in each well</li>
</ul>
</li>
<li><code>readings/*.csv</code>: assay plate readings<ul>
<li>header: machine type, file type ("design" or "readings"), staff ID</li>
<li>blank line</li>
<li>table with column and row titles showing reading from each well</li>
</ul>
</li>
</ul>
</li>
<li>To simulate the messiness of real experimental data,
    the tidy assay plate files in <code>readings/*.csv</code> are copied to <code>mangled/*.csv</code>
    with random changes:<ul>
<li>Some files have a staff member's name added in the first row.</li>
<li>Some have an extra header row containing the experiment date.</li>
<li>Some have a footer with the staff member's ID.</li>
<li>In some, the values are offset one column to the right.</li>
</ul>
</li>
</ul>
<h2 id="workflow">Workflow</h2>
<p>The workflow used to generate the database and data files is shown below:</p>
<ul>
<li><code>snailz</code> or <code>snailz --help</code>: show available commands</li>
<li><code>snailz all</code>: make all datasets</li>
<li><code>snailz map</code>: generate SVG map of sample locations (in progress)</li>
<li><code>snailz mangle</code>: create mangled plate reading files</li>
<li><code>snailz db</code>: generate database</li>
<li><code>snailz plates</code>: generate plate files</li>
<li><code>snailz assays</code>: generate assay files</li>
<li><code>snailz samples</code>: sample snails from survey sites</li>
<li><code>snailz genomes</code>: synthesize genomes</li>
<li><code>snailz grids</code>: synthesize pollution grids</li>
<li><code>snailz clean</code>: remove all datasets</li>
</ul>
<p><img src="https://raw.githubusercontent.com/gvwilson/snailz/main/img/workflow.svg" alt="data generation workflow"></p>
<h2 id="parameters">Parameters</h2>
<p><code>./snailz/params</code> contains the parameter files used to control generation of the reference dataset.
These are included in the package and can be copied into the current directory using <code>snailz params --outdir .</code>
(replace <code>.</code> with another directory name as desired).
<code>snailz params</code> also copies a Makefile that can re-run commands with appropriate parameters;
see the table of commands given earlier for options.</p>
<ul>
<li>Sites: <code>sites.csv</code><ul>
<li><code>site_id</code>: unique label for site (text)</li>
<li><code>lon</code>: longitude of site reference marker (deg)</li>
<li><code>lat</code>: latitude of site reference marker (deg)</li>
</ul>
</li>
<li>Surveys: <code>surveys.csv</code><ul>
<li><code>survey_id</code>: unique label for survey (text)</li>
<li><code>site_id</code>: ID of site where survey was conducted (text)</li>
<li><code>date</code>: date that survey was conducted (date, YYYY-MM-DD)</li>
<li><code>spacing</code>: spacing of measurement point (float, meters)</li>
</ul>
</li>
<li>Genomes: <code>genomes.json</code><ul>
<li><code>length</code>: number of base pairs in sequences (int &gt; 0)</li>
<li><code>num_genomes</code>: how many individuals to generate (int &gt; 0)</li>
<li><code>num_snp</code>: number of single nucleotide polymorphisms (int &gt; 0)</li>
<li><code>prob_other</code>: probability of non-significant mutations (float in 0..1)</li>
<li><code>seed</code>: RNG seed (int &gt; 0)</li>
<li><code>snp_probs</code>: probability of selecting various bases (list of 4 float summing to 1.0)</li>
</ul>
</li>
<li>Grids: <code>grids.json</code><ul>
<li><code>depth</code>: range of random values per cell (int &gt; 0)</li>
<li><code>height</code>: number of cells on Y axis (int &gt; 0)</li>
<li><code>seed</code>: RNG seed (int &gt; 0)</li>
<li><code>width</code>: number of cells on X axis (int &gt; 0)</li>
</ul>
</li>
<li>Assays: <code>assays.json</code><ul>
<li><code>assay_duration</code>: range of days for each assay (ordered pair of int &gt;= 0)</li>
<li><code>assay_plates</code>: range of plates per assay (ordered pair of int &gt;= 1)</li>
<li><code>assay_staff</code>: range of staff in each assay (ordered pair of int &gt; 0)</li>
<li><code>assay_types</code>: types of assays (list of text)</li>
<li><code>control_val</code>: nominal reading value for control wells (float &gt; 0)</li>
<li><code>controls</code>: labels to used for control wells (list of text)</li>
<li><code>enddate</code>: end of all experiments</li>
<li><code>filename_length</code>: length of stem of design/readings filenames (int &gt; 0)</li>
<li><code>invalid</code>: probability of plate being invalidted (float in 0..1)</li>
<li><code>locale</code>: locale to use when generating staff names (text)</li>
<li><code>seed</code>: RNG seed (int &gt; 0)</li>
<li><code>staff</code>: number of staff (int &gt; 0)</li>
<li><code>startdate</code>: start of all experiments</li>
<li><code>stdev</code>: standard deviation on readings (float &gt; 0)</li>
<li><code>treated_val</code>: nominal reading value for treated well (float &gt; 0)</li>
<li><code>treatment</code>: label to use for treated wells (text)</li>
</ul>
</li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": ".", "features": [], "search": "assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="assets/javascripts/bundle.fe8b6f2b.min.js"></script>
      
    
  </body>
</html>